src/cli/self_update/windows.rs

```diff
@@ -41,10 +48,102 @@ pub(crate) fn do_msvc_check(opts: &InstallOpts<'_>) -> bool {
     let installing_msvc = host_triple.contains("msvc");
     let have_msvc = windows_registry::find_tool(&host_triple, "cl.exe").is_some();
     if installing_msvc && !have_msvc {
-        return false;
+        // Visual Studio build tools are required.
+        // If the user does not have Visual Studio installed and their host
+        // machine is i686 or x86_64 then it's OK to try an auto install.
+        // Otherwise a manual install will be required.
+        let has_any_vs = windows_registry::find_vs_version().is_ok();
+        let is_x86 = host_triple.contains("i686") || host_triple.contains("x86_64");
+        if is_x86 && !has_any_vs {
+            Some(VsInstallPlan::Automatic)
+        } else {
+            Some(VsInstallPlan::Manual)
+        }
+    } else {
+        None
+    }
+}
+
+#[derive(Debug)]
+struct VsInstallError(i32);
+impl std::error::Error for VsInstallError {}
+impl fmt::Display for VsInstallError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // See https://docs.microsoft.com/en-us/visualstudio/install/use-command-line-parameters-to-install-visual-studio?view=vs-2022#error-codes
+        let message = match self.0 {
+            740 => "elevation required",
+            1001 => "Visual Studio installer process is running",
+            1003 => "Visual Studio is in use",
+            1602 => "operation was canceled",
+            1618 => "another installation running",
+            1641 => "operation completed successfully, and reboot was initiated",
+            3010 => "operation completed successfully, but install requires reboot before it can be used",
+            5003 => "bootstrapper failed to download installer",
+            5004 => "operation was canceled",
+            5005 => "bootstrapper command-line parse error",
+            5007 => "operation was blocked - the computer does not meet the requirements",
+            8001 => "arm machine check failure",
+            8002 => "background download precheck failure",
+            8003 => "out of support selectable failure",
+            8004 => "target directory failure",
+            8005 => "verifying source payloads failure",
+            8006 => "Visual Studio processes running",
+            -1073720687 => "connectivity failure",
+            -1073741510 => "Microsoft Visual Studio Installer was terminated",
+            _ => "error installing Visual Studio"
+        };
+        write!(f, "{} (exit code {})", message, self.0)
     }
+}
 
-    true
+pub(crate) fn try_install_msvc() -> Result<()> {
+    // download the installer
+    let visual_studio_url = utils::parse_url("https://aka.ms/vs/17/release/vs_community.exe")?;
+
+    let tempdir = tempfile::Builder::new()
+        .prefix("rustup-visualstudio")
+        .tempdir()
+        .context("error creating temp directory")?;
+
+    let visual_studio = tempdir.path().join("vs_setup.exe");
+    utils::download_file(&visual_studio_url, &visual_studio, None, &|_| ())?;
```

Comment:
It's a pity you don't offer a download progress bar here. I accept it might be a bit of a faff, but if you can sort a notification handler out to get a progress bar, that'd be nice for people.

Reply:
That would indeed be good. I just have to learn how 🙂. It should only be a small download though (about 1.6 MB) as it bootstraps the actual install.

Reply:
It's not a blocker, just if you want to have a go at it you're welcome to :D

Reply:
Yep, I can definitely try!

Reply:
Ok, I've added a progress bar.

Reply:
That looks quite promising, I've restarted the CI to clear that ARM problem.